pdns/Makefile.am

```diff
@@ -1663,8 +1663,9 @@ dnsdist:
 if HAVE_SYSTEMD
 pdns.service: pdns.service.in
 	$(AM_V_GEN)sed -e 's![@]sbindir[@]!$(sbindir)!' -e 's![@]service_user[@]!$(service_user)!' -e 's![@]service_group[@]!$(service_group)!' < $< > $@
-if !HAVE_SYSTEMD_WITH_RUNTIME_DIR_ENV
+if HAVE_SYSTEMD_WITH_RUNTIME_DIR_ENV
 	$(AM_V_GEN)sed -e 's!/pdns_server!& --socket-dir=%t/pdns!' -i $@
+	$(AM_V_GEN)perl -ni -e 'print unless /^RuntimeDirectory/' $@
```

Comment:
I'm late to the party so I'm very likely missing something, but doesn't that change actually removes the `RuntimeDirectory` line when `RuntimeDirectory` is supported?

Reply:
`RuntimeDirectory` supported != Support for the `RUNTIME_DIR` env var :(

Reply:
Right, so I get that the changes in the spec files add `RuntimeDirectory` back to the unit file on EL 7 and 8. I also understand this change here only adds `--socket-dir=%t/pdns` when `systemd >= 240` which seems good. What I don't understand is that it looks like it will remove the `RuntimeDirectory` line when `systemd >= 240`, and leave it otherwise?

Reply:
because we can't detect what the runtime dir is, so we fall back to the compiled in `socket-dir` (/var/run) usually

Reply:
(https://github.com/PowerDNS/pdns/blob/d606317b3ab6e0eff8537373d83d609c1745ef11/pdns/common_startup.cc#L80-L86)

Reply:
So am I correctly understanding that the final states are:

Before systemd 240 (RuntimeDirectory environment variable not supported):
- no socket-dir=%t so default to LOCALSTATEDIR/pdns
- RuntimeDirectory is set

Except on EL 7 and 8 where:
- socket-dir=%t is set, althought it will likely get overriden by the RUNTIME_DIRECTORY env variable?
- RuntimeDirectory is set back (so it's now there twice?)

After 240:
- socket-dir=%t is set
- RuntimeDirectory is removed

I only tested the last one (on Arch).

Reply:
This is produced `pdns-recursor@.service` on a centos-7 build:
```
[Unit]
Description=PowerDNS Recursor %i
Documentation=man:pdns_recursor(1) man:rec_control(1)
Documentation=https://doc.powerdns.com
Wants=network-online.target nss-lookup.target
Before=nss-lookup.target
After=network-online.target

[Service]
ExecStart=/usr/sbin/pdns_recursor --socket-dir=%t/pdns-recursor-%i --config-name=%i --daemon=no --write-pid=no --disab
le-syslog --log-timestamp=no
User=pdns-recursor
Group=pdns-recursor
Type=notify
Restart=on-failure
StartLimitInterval=0
RuntimeDirectory=pdns-recursor-%i
SyslogIdentifier=pdns-recursor-%i

# Tuning
LimitNOFILE=16384

RuntimeDirectory=pdns-recursor-%i
# Sandboxing
CapabilityBoundingSet=CAP_NET_BIND_SERVICE CAP_CHOWN
AmbientCapabilities=CAP_NET_BIND_SERVICE CAP_CHOWN
NoNewPrivileges=true
PrivateDevices=true
PrivateTmp=true
# Setting PrivateUsers=true prevents us from opening our sockets
ProtectHome=true
ProtectSystem=full
RestrictAddressFamilies=AF_UNIX AF_INET AF_INET6
SystemCallArchitectures=native

[Install]
WantedBy=multi-user.target
```
Indeed, two `RuntimeDirectory` lines.

Reply:
Ha! That is wrong indeed.

it should be:

* before 240
   * no `socket-dir=%t` so default to `LOCALSTATEDIR/pdns`
   * RuntimeDirectory set to `LOCALSTATEDIR/pdns` (or kept at 'pdns', as that is the default) to ensure the dir exists
* after 240
   * `socket-dir=%t` is set and modified for the @.service
   * `RuntimeDirectory` is set (and modified for the @.service